rerun_cpp/src/rerun/component_batch.hpp

```diff
@@ -1,94 +1,336 @@
+    /// outlives the component batch. Owned component batches on the other hand take ownership by
+    /// allocating a std::vector and moving the data into it. This is typically only required when
+    /// passing in temporary objects into an adapter or non-trivial data conversion is necessary.
+    ///
+    /// By implementing your own adapters for certain component types, you can map your data to
+    /// Rerun types which then can be logged.
+    ///
+    /// To implement an adapter for a type T, specialize `ComponentBatchAdapter<TComponent, T>` and
+    /// define `ComponentBatch<TComponent> operator()(const T& input)`.
+    /// It is *highly recommended* to also specify `ComponentBatch<TComponent> operator()(T&&
+    /// input)` in order to to accidentally borrow data that is passed in as a temporary!
+    ///
+    /// TODO(andreas): Point to an example here and in the assert.
+    template <typename TComponent, typename TInput>
+    struct ComponentBatchAdapter {
+        template <typename... Ts>
+        struct NoAdapterFor : std::false_type {};
+
+        static_assert(
+            NoAdapterFor<TComponent, TInput>::value, // Always evaluate to false, but in a way that
+                                                     // requires template instantiation.
+            "ComponentBatchAdapter is not implemented for this type. "
+            "It is implemented for for single components as well as std::vector, std::array, and "
+            "c-arrays of components. "
+            "You can add your own implementation by specializing "
+            "ComponentBatchAdapter<TComponent, T> for a given "
+            "component and your input type T."
+        );
+    };
+
+    /// Type of ownership of the the batch's data.
+    ///
+    /// User access to this is typically only needed for debugging and testing.
+    enum class BatchOwnership {
+        /// The component batch does not own the data and only has a pointer and a
+        /// size.
+        Borrowed,
+
+        /// The component batch owns the data via an std::vector.
+        VectorOwned,
+    };
+
     /// Generic list of components that are contiguous in memory.
     ///
-    /// Does *not* own the data, user is responsible for the lifetime independent of how it was
-    /// passed in.
-    template <typename ComponentType>
+    /// Any creation from a non-temporary will neither copy the data nor take ownership of it.
+    /// This means that any data passed in (unless temporary) must outlive the component batch!
+    ///
+    /// However, when created from a temporary, the component batch will take ownership of the data.
+    /// For details, refer to the documentation of the respective constructor.
+    ///
+    /// Implementation notes:
+    ///
+    /// Does intentionally not implement copy construction since this for the owned case this may
+    /// be expensive.Typically, there should be no need to copy component batches, so this more than
```

Comment:
```suggestion
    /// ## Implementation notes
    ///
    /// Does intentionally not implement copy construction since this for the owned case this may
    /// be expensive. Typically, there should be no need to copy component batches, so this more than
```